### ISSUE TITLE
checkIfNeedEnqueue returns FutureOr while _dispatchRequest expects Future

### ISSUE THREAD

**Issue by eelishiram** [flutterchina/dio#783]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |  android    |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |   1.17.0   |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |  3.0.9    |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |  100%    |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Full stacktrace:**
```
I/flutter (30164): intercept onError : DioError [DioErrorType.DEFAULT]: type 'Future<dynamic>' is not a subtype of type 'FutureOr<Response<dynamic>>'
I/flutter (30164): #0      DioMixin._dispatchRequest (package:dio/src/dio.dart:954:9)
I/flutter (30164): <asynchronous suspension>
I/flutter (30164): #1      DioMixin._request._interceptorWrapper.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:dio/src/dio.dart:849:37)
I/flutter (30164): #2      DioMixin.checkIfNeedEnqueue (package:dio/src/dio.dart:1117:22)
I/flutter (30164): #3      DioMixin._request._interceptorWrapper.<anonymous closure>.<anonymous closure> (package:dio/src/dio.dart:846:22)
I/flutter (30164): #4      new Future.<anonymous closure> (dart:async/future.dart:176:37)
I/flutter (30164): #5      _rootRun (dart:async/zone.dart:1180:38)
I/flutter (30164): #6      _CustomZone.run (dart:async/zone.dart:1077:19)
I/flutter (30164): #7      _CustomZone.runGuarded (dart:async/zone.dart:979:7)
I/flutter (30164): #8      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1019:23)
I/flutter (30164): #9      _rootRun (dart:async/zone.dart:1184:13)
I/flutter (30164): #10     _CustomZone.run (dart:async/zone.dart:1077:19)
I/flutter (30164): #11     _CustomZone.bindCallback.<anonym

```
After upgrading to Flutter v1.17.0, I experience this exception 
`intercept onError : DioError [DioErrorType.DEFAULT]: type 'Future<dynamic>' is not a subtype of type 'FutureOr<Response<dynamic>>'`
Upon checking the source code, I saw that **checkIfNeedEnqueue** returns **FutureOr** from the body of  **_dispatchRequest**.  Is there something that I am missing?

**Comment by ARDcode** [comment]
I have this issue too with POST requests =(

**Comment by supermebing** [comment]
问题根本没有解决

**Comment by iamnabink** [comment]
The issue is still persisting, any solution for this?

**Comment by diegodalbosco** [comment]
I spent an entire day to figure out a solution to this problem. In my case I was using an onError interceptor that was an async function returning a Future, I resolved by making this function synchronous and returning void. Inside I used .then/.catch instead of async/await. I'm not sure why this worked for me..
